File: example/web/browser-redirect.js
/**
 * Browser Redirect Script for PWA Mobile Install
 * 
 * Detects in-app browsers (Instagram, Facebook, LinkedIn, etc.) and shows
 * an overlay with instructions to open in the default browser.
 * 
 * USAGE:
 * 1. Copy this file and browser-redirect-config.js to your web/ directory
 * 2. Add to your index.html <head> section:
 *    <script src="browser-redirect-config.js"></script>
 *    <script src="browser-redirect.js"></script>
 */

(function() {
  'use strict';

  /**
   * Detects if the current browser is an in-app browser
   * @returns {Object} Detection result with browser name and whether it's in-app
   */
  function detectInAppBrowser() {
    const ua = navigator.userAgent || navigator.vendor || window.opera;
    
    // Instagram in-app browser
    if (ua.indexOf('Instagram') > -1) {
      return { isInApp: true, browser: 'Instagram' };
    }
    
    // Facebook/Messenger in-app browser (FBAN = Facebook App, FBAV = Facebook App Version)
    if (ua.indexOf('FBAN') > -1 || ua.indexOf('FBAV') > -1) {
      return { isInApp: true, browser: 'Facebook' };
    }
    
    // LinkedIn in-app browser
    if (ua.indexOf('LinkedInApp') > -1) {
      return { isInApp: true, browser: 'LinkedIn' };
    }
    
    // Twitter in-app browser
    if (ua.indexOf('Twitter') > -1) {
      return { isInApp: true, browser: 'Twitter' };
    }
    
    // TikTok in-app browser
    if (ua.indexOf('musical_ly') > -1 || ua.indexOf('TikTok') > -1) {
      return { isInApp: true, browser: 'TikTok' };
    }
    
    // Snapchat in-app browser
    if (ua.indexOf('Snapchat') > -1) {
      return { isInApp: true, browser: 'Snapchat' };
    }
    
    // WhatsApp in-app browser
    if (ua.indexOf('WhatsApp') > -1) {
      return { isInApp: true, browser: 'WhatsApp' };
    }
    
    // Line in-app browser
    if (ua.indexOf('Line') > -1) {
      return { isInApp: true, browser: 'Line' };
    }
    
    // WeChat in-app browser
    if (ua.indexOf('MicroMessenger') > -1) {
      return { isInApp: true, browser: 'WeChat' };
    }
    
    return { isInApp: false, browser: 'Standard' };
  }

  /**
   * Shows an overlay with instructions to open in default browser
   * @param {string} browserName Name of the detected in-app browser
   */
  function showBrowserRedirectOverlay(browserName) {
    // Check if overlay already exists
    if (document.getElementById('browser-redirect-overlay')) {
      return;
    }

    const config = window.BROWSER_REDIRECT_CONFIG || {};
    const colors = config.colors || {};
    const messages = config.messages || {};

    const primaryColor = colors.primary || '#1A1D32';
    const accentColor = colors.accent || '#FF7A5A';
    const surfaceColor = colors.surface || '#1E1E1E';
    const textColor = colors.text || '#FFFFFF';

    const overlay = document.createElement('div');
    overlay.id = 'browser-redirect-overlay';
    overlay.innerHTML = `
      <style>
        #browser-redirect-overlay {
          position: fixed;
          top: 0;
          left: 0;
          right: 0;
          bottom: 0;
          background: linear-gradient(135deg, ${primaryColor} 0%, #725D78 100%);
          z-index: 999999;
          display: flex;
          align-items: center;
          justify-content: center;
          padding: 24px;
          font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif;
          animation: fadeIn 0.3s ease-out;
        }
        
        @keyframes fadeIn {
          from { opacity: 0; }
          to { opacity: 1; }
        }
        
        .redirect-content {
          background: ${surfaceColor};
          border-radius: 16px;
          padding: 32px 24px;
          max-width: 400px;
          width: 100%;
          text-align: center;
          animation: slideUp 0.4s ease-out;
        }
        
        @keyframes slideUp {
          from {
            transform: translateY(30px);
            opacity: 0;
          }
          to {
            transform: translateY(0);
            opacity: 1;
          }
        }
        
        .redirect-icon {
          width: 80px;
          height: 80px;
          margin: 0 auto 24px;
          padding: 20px;
          background: rgba(255, 255, 255, 0.1);
          border-radius: 50%;
          display: flex;
          align-items: center;
          justify-content: center;
          font-size: 40px;
        }
        
        .redirect-title {
          font-size: 24px;
          font-weight: 700;
          margin-bottom: 16px;
        }
        
        .redirect-title-accent { color: ${accentColor}; }
        .redirect-title-main { color: ${textColor}; }
        
        .redirect-message {
          font-size: 16px;
          color: ${textColor};
          line-height: 1.6;
          margin-bottom: 32px;
        }
        
        .redirect-steps {
          background: rgba(255, 255, 255, 0.05);
          border-radius: 12px;
          margin-bottom: 24px;
          text-align: left;
        }
        
        .redirect-step {
          display: flex;
          align-items: center;
          justify-content: space-between;
          padding: 16px;
          font-size: 14px;
          color: ${textColor};
          border-bottom: 1px solid rgba(255, 255, 255, 0.1);
        }
        
        .redirect-step:last-child {
          border-bottom: none;
        }
        
        .step-icon {
          width: 36px;
          height: 36px;
          background: rgba(${hexToRgb(accentColor)}, 0.3);
          border-radius: 8px;
          display: flex;
          align-items: center;
          justify-content: center;
          flex-shrink: 0;
          margin-left: 12px;
          color: ${accentColor};
          font-size: 18px;
        }
        
        .redirect-button {
          background: rgba(${hexToRgb(accentColor)}, 0.3);
          color: ${accentColor};
          border: none;
          border-radius: 12px;
          padding: 16px 28px;
          font-size: 16px;
          font-weight: 600;
          cursor: pointer;
          width: 100%;
          transition: all 0.3s ease;
          display: flex;
          align-items: center;
          justify-content: center;
          gap: 8px;
        }
        
        .redirect-button:hover {
          background: rgba(${hexToRgb(accentColor)}, 0.4);
          transform: translateY(-2px);
        }
        
        .redirect-button:active {
          transform: translateY(0);
        }
        
        .redirect-dismiss {
          background: transparent;
          color: rgba(255, 255, 255, 0.6);
          border: none;
          padding: 12px;
          font-size: 14px;
          cursor: pointer;
          margin-top: 12px;
        }
      </style>
      
      <div class="redirect-content">
        <div class="redirect-icon">🌐</div>
        <div class="redirect-title">
          <span class="redirect-title-accent">${messages.titleAccent || 'Open in'}</span>
          <span class="redirect-title-main">${messages.titleMain || ' Browser'}</span>
        </div>
        <div class="redirect-message">
          ${messages.message || 'For the best experience and to enable all features, please open this link in your default browser.'}
        </div>
        
        <div class="redirect-steps">
          ${getInstructionsForBrowser(browserName)}
        </div>
        
        <button class="redirect-button" onclick="attemptBrowserRedirect()">
          <span>${isIOS() ? '🧭' : '🌐'}</span>
          <span>${isIOS() ? 'Open in Safari' : 'Open in Chrome'}</span>
        </button>
        
        ${config.allowDismiss !== false ? '<button class="redirect-dismiss" onclick="dismissRedirectOverlay()">Continue anyway</button>' : ''}
      </div>
    `;
    
    document.body.appendChild(overlay);
  }

  /**
   * Helper to convert hex color to RGB
   */
  function hexToRgb(hex) {
    const result = /^#?([a-f\d]{2})([a-f\d]{2})([a-f\d]{2})$/i.exec(hex);
    return result ? 
      `${parseInt(result[1], 16)}, ${parseInt(result[2], 16)}, ${parseInt(result[3], 16)}` : 
      '255, 122, 90';
  }

  /**
   * Detect iOS device
   */
  function isIOS() {
    return /iPad|iPhone|iPod/.test(navigator.userAgent);
  }

  /**
   * Gets platform-specific instructions for opening in default browser
   * @param {string} browserName Name of the detected in-app browser
   * @returns {string} HTML string with instructions
   */
  function getInstructionsForBrowser(browserName) {
    const iOS = isIOS();
    const isAndroid = /Android/.test(navigator.userAgent);
    
    let instructions = '';
    
    if (browserName === 'Instagram') {
      if (iOS) {
        instructions = `
          <div class="redirect-step">
            <div>Tap the three dots (•••) in the top right corner</div>
            <div class="step-icon">⋯</div>
          </div>
          <div class="redirect-step">
            <div>Select "Open in Safari"</div>
            <div class="step-icon">🧭</div>
          </div>
        `;
      } else if (isAndroid) {
        instructions = `
          <div class="redirect-step">
            <div>Tap the three dots (⋮) in the top right corner</div>
            <div class="step-icon">⋮</div>
          </div>
          <div class="redirect-step">
            <div>Select "Open in Chrome"</div>
            <div class="step-icon">🌐</div>
          </div>
        `;
      }
    } else if (browserName === 'Facebook') {
      if (iOS) {
        instructions = `
          <div class="redirect-step">
            <div>Tap the three dots (•••) in the bottom right</div>
            <div class="step-icon">⋯</div>
          </div>
          <div class="redirect-step">
            <div>Select "Open in Safari"</div>
            <div class="step-icon">🧭</div>
          </div>
        `;
      } else if (isAndroid) {
        instructions = `
          <div class="redirect-step">
            <div>Tap the three dots (⋮) in the top right corner</div>
            <div class="step-icon">⋮</div>
          </div>
          <div class="redirect-step">
            <div>Select "Open in external browser"</div>
            <div class="step-icon">🌐</div>
          </div>
        `;
      }
    } else if (browserName === 'LinkedIn') {
      instructions = `
        <div class="redirect-step">
          <div>Tap the three dots (⋮) or share icon</div>
          <div class="step-icon">⋮</div>
        </div>
        <div class="redirect-step">
          <div>Select "Open in ${iOS ? 'Safari' : 'Chrome'}"</div>
          <div class="step-icon">${iOS ? '🧭' : '🌐'}</div>
        </div>
      `;
    } else {
      // Generic instructions
      instructions = `
        <div class="redirect-step">
          <div>Look for a menu icon (⋮ or •••) in the app</div>
          <div class="step-icon">⋮</div>
        </div>
        <div class="redirect-step">
          <div>Select "Open in ${iOS ? 'Safari' : 'Chrome'}"</div>
          <div class="step-icon">${iOS ? '🧭' : '🌐'}</div>
        </div>
      `;
    }
    
    return instructions;
  }

  /**
   * Attempts to open the current URL in the default browser
   */
  window.attemptBrowserRedirect = function() {
    const currentUrl = window.location.href;
    const iOS = isIOS();
    const isAndroid = /Android/.test(navigator.userAgent);
    
    if (isAndroid) {
      // Android: Try Chrome Intent URL
      const host = window.location.host;
      const path = window.location.pathname + window.location.search + window.location.hash;
      const chromeIntent = `intent://${host}${path}#Intent;scheme=https;package=com.android.chrome;end`;
      
      window.location.href = chromeIntent;
      
      // Fallback after delay
      setTimeout(() => {
        const genericIntent = `intent://${host}${path}#Intent;scheme=https;action=android.intent.action.VIEW;end`;
        window.location.href = genericIntent;
      }, 1000);
      
    } else if (iOS) {
      // iOS: Try window.open with _system target
      window.open(currentUrl, '_system');
    }
  };

  /**
   * Dismisses the redirect overlay
   */
  window.dismissRedirectOverlay = function() {
    const overlay = document.getElementById('browser-redirect-overlay');
    if (overlay) {
      overlay.style.animation = 'fadeOut 0.3s ease-out';
      setTimeout(() => overlay.remove(), 300);
    }
  };

  /**
   * Initialize the browser detection and redirect logic
   */
  function init() {
    const config = window.BROWSER_REDIRECT_CONFIG || {};
    
    // Check if feature is enabled
    if (config.enabled === false) {
      return;
    }
    
    // Check if current path is excluded
    const currentPath = window.location.pathname;
    if (config.excludePaths && config.excludePaths.some(path => currentPath.startsWith(path))) {
      return;
    }
    
    const detection = detectInAppBrowser();
    
    if (detection.isInApp) {
      console.log(`[PWA Install] Detected in-app browser: ${detection.browser}`);
      
      // Show overlay with instructions (if enabled)
      if (config.showOverlay !== false) {
        const delay = config.overlayDelay || 0;
        setTimeout(() => showBrowserRedirectOverlay(detection.browser), delay);
      }
      
      // Log analytics event if available and enabled
      if (config.analytics !== false && window.gtag) {
        const eventName = (config.analytics && config.analytics.eventName) || 'in_app_browser_detected';
        window.gtag('event', eventName, {
          browser: detection.browser,
          user_agent: navigator.userAgent
        });
      }
    }
  }

  // Run detection when DOM is ready
  if (document.readyState === 'loading') {
    document.addEventListener('DOMContentLoaded', init);
  } else {
    init();
  }
})();

File: lib/src/core/browser_redirect.dart
import 'package:flutter/foundation.dart';
import 'package:web/web.dart' as web;

/// Injects a JavaScript-based in-app browser detection and redirect overlay.
///
/// This class provides a static method to inject a script that:
/// - Detects in-app browsers (Instagram, Facebook, TikTok, etc.)
/// - Shows an overlay with instructions to open in the default browser
/// - Provides auto-redirect buttons for iOS Safari and Android Chrome
///
/// The script executes immediately when injected, BEFORE Flutter fully loads.
class BrowserRedirectInjector {
  static bool _injected = false;

  /// Injects the browser redirect script into the page.
  ///
  /// This method is idempotent - calling it multiple times has no effect
  /// after the first call.
  ///
  /// [nonce] - Optional CSP nonce for apps using strict Content Security Policy.
  /// If your app has a CSP header with `script-src` that doesn't include
  /// `'unsafe-inline'`, pass the nonce value here to allow the script to execute.
  static void inject({String? nonce}) {
    if (!kIsWeb || _injected) return;

    try {
      final script =
          web.document.createElement('script') as web.HTMLScriptElement;
      script.text = _getRedirectScript();
      if (nonce != null) {
        script.setAttribute('nonce', nonce);
      }
      web.document.head?.appendChild(script);
      _injected = true;
      debugPrint('BrowserRedirectInjector: Script injected');
    } catch (e) {
      debugPrint('BrowserRedirectInjector: Failed to inject script: $e');
    }
  }

  /// Checks if the redirect overlay is currently shown.
  static bool isOverlayShown() {
    if (!kIsWeb) return false;

    try {
      final overlay =
          web.document.getElementById('pwa-browser-redirect-overlay');
      return overlay != null;
    } catch (e) {
      return false;
    }
  }

  /// Removes the redirect overlay from the DOM if it exists.
  static void removeOverlay() {
    if (!kIsWeb) return;

    try {
      final overlay =
          web.document.getElementById('pwa-browser-redirect-overlay');
      overlay?.remove();
    } catch (e) {
      debugPrint('BrowserRedirectInjector: Failed to remove overlay: $e');
    }
  }

  static String _getRedirectScript() {
    const accentColor = '#FF7A5A';
    const backgroundColor = '#1A1D32';
    const surfaceColor = '#1E1E1E';
    const textColor = '#FFFFFF';
    const title = 'Open in Browser';
    const message =
        'For the best experience and to enable all features, please open this link in your default browser.';

    // The JavaScript below is organized into sections:
    // 1. PWA Detection - Skip if already running as installed PWA
    // 2. In-App Browser Detection - Identify Instagram, Facebook, TikTok, etc.
    // 3. Overlay UI Generation - Create the redirect instruction overlay
    // 4. Redirect Logic - Attempt to open in native browser with error handling
    // 5. Clipboard Handling - Copy link functionality with fallback
    // 6. Initialization - Detect and show overlay if in-app browser
    return '''
(function() {
  'use strict';
  
  // ============================================================
  // SECTION 1: PWA Detection
  // Skip the entire script if already running as an installed PWA
  // ============================================================
  if (window.matchMedia('(display-mode: standalone)').matches ||
      window.matchMedia('(display-mode: fullscreen)').matches ||
      window.navigator.standalone === true) {
    return;
  }

  // ============================================================
  // SECTION 2: In-App Browser Detection
  // Checks user agent for known in-app browser signatures
  // Returns: { isInApp: boolean, browser: string }
  // ============================================================
  function detectInAppBrowser() {
    var ua = (navigator.userAgent || navigator.vendor || window.opera || '').toLowerCase();
    
    // Social media apps - each has unique UA tokens
    if (ua.indexOf('instagram') > -1) return { isInApp: true, browser: 'Instagram' };
    if (ua.indexOf('fban') > -1 || ua.indexOf('fbav') > -1 || ua.indexOf('messenger') > -1) return { isInApp: true, browser: 'Facebook' };
    if (ua.indexOf('linkedinapp') > -1) return { isInApp: true, browser: 'LinkedIn' };
    if (ua.indexOf('twitter') > -1) return { isInApp: true, browser: 'Twitter' };
    if (ua.indexOf('musical_ly') > -1 || ua.indexOf('tiktok') > -1 || ua.indexOf('bytedance') > -1) return { isInApp: true, browser: 'TikTok' };
    if (ua.indexOf('snapchat') > -1) return { isInApp: true, browser: 'Snapchat' };
    if (ua.indexOf('whatsapp') > -1) return { isInApp: true, browser: 'WhatsApp' };
    if (ua.indexOf('line/') > -1) return { isInApp: true, browser: 'Line' };
    if (ua.indexOf('micromessenger') > -1) return { isInApp: true, browser: 'WeChat' };
    if (ua.indexOf('slack') > -1) return { isInApp: true, browser: 'Slack' };
    if (ua.indexOf('discord') > -1) return { isInApp: true, browser: 'Discord' };
    
    // Generic Android WebView detection ("; wv)" token in UA)
    var isAndroid = ua.indexOf('android') > -1;
    var hasWebViewToken = ua.indexOf('; wv)') > -1 || ua.indexOf(';wv)') > -1;
    if (isAndroid && hasWebViewToken) return { isInApp: true, browser: 'In-App Browser' };
    
    return { isInApp: false, browser: 'Standard' };
  }

  // ============================================================
  // SECTION 3: Overlay UI Generation
  // Creates the visual overlay with instructions for opening in browser
  // Instructions are customized based on detected browser and platform
  // ============================================================
  function showOverlay(browserName) {
    if (document.getElementById('pwa-browser-redirect-overlay')) return;

    var isIOS = /ipad|iphone|ipod/.test(navigator.userAgent.toLowerCase());
    var step1Text, step2Text;
    
    // Customize instructions based on the detected in-app browser
    if (browserName === 'Instagram') {
      step1Text = isIOS ? 'Tap the three dots (•••) in the top right corner' : 'Tap the three dots (⋮) in the top right corner';
      step2Text = isIOS ? 'Select "Open in Safari"' : 'Select "Open in Chrome"';
    } else if (browserName === 'Facebook') {
      step1Text = isIOS ? 'Tap the three dots (•••) in the bottom right corner' : 'Tap the three dots (⋮) in the top right corner';
      step2Text = isIOS ? 'Select "Open in Safari"' : 'Select "Open in external browser"';
    } else if (browserName === 'TikTok') {
      step1Text = 'Tap the three dots (•••) or share button';
      step2Text = isIOS ? 'Select "Open in Safari"' : 'Select "Open in browser"';
    } else if (browserName === 'Discord') {
      step1Text = 'Tap the three dots (•••) or share';
      step2Text = 'Select "Open in Browser"';
    } else {
      step1Text = 'Look for a menu icon (⋮ or •••) in the app';
      step2Text = isIOS ? 'Select "Open in Safari"' : 'Select "Open in Chrome"';
    }

    var overlay = document.createElement('div');
    overlay.id = 'pwa-browser-redirect-overlay';
    overlay.innerHTML = [
      '<style>',
      '#pwa-browser-redirect-overlay { position: fixed; top: 0; left: 0; right: 0; bottom: 0; background: $backgroundColor; z-index: 999999; display: flex; align-items: center; justify-content: center; padding: 24px; font-family: -apple-system, BlinkMacSystemFont, "Segoe UI", Roboto, sans-serif; animation: pwaFadeIn 0.3s ease-out; }',
      '@keyframes pwaFadeIn { from { opacity: 0; } to { opacity: 1; } }',
      '.pwa-redirect-content { background: $surfaceColor; border-radius: 16px; padding: 32px 24px; max-width: 400px; width: 100%; text-align: center; }',
      '.pwa-redirect-icon { width: 80px; height: 80px; margin: 0 auto 24px; background: rgba(255,255,255,0.1); border-radius: 50%; display: flex; align-items: center; justify-content: center; font-size: 40px; }',
      '.pwa-redirect-title { font-size: 24px; font-weight: 700; margin-bottom: 16px; color: $textColor; }',
      '.pwa-redirect-title span { color: $accentColor; }',
      '.pwa-redirect-message { font-size: 16px; color: $textColor; line-height: 1.6; margin-bottom: 32px; opacity: 0.9; }',
      '.pwa-redirect-step { display: flex; align-items: center; justify-content: space-between; padding: 16px; background: rgba(255,255,255,0.05); border-radius: 12px; margin-bottom: 12px; color: $textColor; font-size: 14px; text-align: left; }',
      '.pwa-step-icon { width: 40px; height: 40px; background: ' + '$accentColor' + '33; border-radius: 8px; display: flex; align-items: center; justify-content: center; color: $accentColor; font-size: 20px; flex-shrink: 0; margin-left: 12px; }',
      '.pwa-redirect-button { background: ' + '$accentColor' + '33; color: $accentColor; border: none; border-radius: 12px; padding: 16px 24px; font-size: 16px; font-weight: 600; cursor: pointer; width: 100%; margin-top: 8px; display: flex; align-items: center; justify-content: center; gap: 8px; transition: opacity 0.2s; }',
      '.pwa-redirect-button:disabled { opacity: 0.6; cursor: not-allowed; }',
      '.pwa-redirect-button.pwa-redirect-failed { background: rgba(255,100,100,0.2); color: #ff6b6b; }',
      '.pwa-copy-button { background: transparent; color: rgba(255,255,255,0.7); border: 1px solid rgba(255,255,255,0.2); border-radius: 12px; padding: 14px 24px; font-size: 14px; cursor: pointer; width: 100%; margin-top: 12px; }',
      '</style>',
      '<div class="pwa-redirect-content">',
      '  <div class="pwa-redirect-icon">🌐</div>',
      '  <div class="pwa-redirect-title"><span>' + '$title' + '</span></div>',
      '  <div class="pwa-redirect-message">$message</div>',
      '  <div class="pwa-redirect-step"><div>' + step1Text + '</div><div class="pwa-step-icon">⋯</div></div>',
      '  <div class="pwa-redirect-step"><div>' + step2Text + '</div><div class="pwa-step-icon">' + (isIOS ? '🧭' : '🌐') + '</div></div>',
      '  <button class="pwa-redirect-button" id="pwa-redirect-btn" onclick="pwaAttemptRedirect()">' + (isIOS ? '🧭 Open in Safari' : '🌐 Open in Chrome') + '</button>',
      '  <button class="pwa-copy-button" onclick="pwaCopyLink()">📋 Copy Link</button>',
      '</div>'
    ].join('\\n');

    document.body.appendChild(overlay);
  }

  // ============================================================
  // SECTION 4: Redirect Logic
  // Attempts to open the current URL in the native browser.
  // Uses intent:// URLs on Android and x-safari-https:// on iOS.
  // Includes timeout-based error detection to show fallback UI.
  // ============================================================
  
  // Track redirect state to detect failures
  var redirectAttempted = false;
  var redirectTimeout = null;
  
  // Shows fallback UI when redirect fails (page still visible after timeout)
  function showRedirectFailed() {
    var btn = document.getElementById('pwa-redirect-btn');
    if (btn && !btn.classList.contains('pwa-redirect-failed')) {
      btn.classList.add('pwa-redirect-failed');
      btn.innerHTML = '⚠️ Auto-open failed - use steps above';
      btn.disabled = true;
      console.log('PWA: Redirect failed, showing fallback UI');
    }
  }

  window.pwaAttemptRedirect = function() {
    var btn = document.getElementById('pwa-redirect-btn');
    var host = window.location.host;
    var path = window.location.pathname + window.location.search + window.location.hash;
    var isIOS = /ipad|iphone|ipod/.test(navigator.userAgent.toLowerCase());
    var isAndroid = /android/.test(navigator.userAgent.toLowerCase());

    // Prevent multiple rapid attempts
    if (redirectAttempted) {
      console.log('PWA: Redirect already attempted');
      return;
    }
    redirectAttempted = true;
    
    // Update button to show attempting state
    if (btn) {
      btn.innerHTML = '⏳ Opening browser...';
      btn.disabled = true;
    }

    // Set timeout to detect if redirect failed (page still visible)
    // If redirect succeeds, this page will be hidden/unloaded
    redirectTimeout = setTimeout(function() {
      // Check if document is still visible (redirect likely failed)
      if (!document.hidden) {
        showRedirectFailed();
      }
    }, 2500);
    
    // Listen for visibility change - if page becomes hidden, redirect worked
    document.addEventListener('visibilitychange', function onVisChange() {
      if (document.hidden && redirectTimeout) {
        clearTimeout(redirectTimeout);
        redirectTimeout = null;
        document.removeEventListener('visibilitychange', onVisChange);
      }
    });

    try {
      if (isAndroid) {
        // Android: Try Chrome-specific intent first, then generic intent
        var chromeIntent = 'intent://' + host + path + '#Intent;scheme=https;package=com.android.chrome;end';
        window.location.href = chromeIntent;
        
        // Fallback to generic browser intent after 1 second
        setTimeout(function() {
          if (!document.hidden) {
            var genericIntent = 'intent://' + host + path + '#Intent;scheme=https;action=android.intent.action.VIEW;end';
            window.location.href = genericIntent;
          }
        }, 1000);
      } else if (isIOS) {
        // iOS: Try x-safari-https scheme, fallback to window.open
        var currentUrl = window.location.href;
        var safariUrl = currentUrl.replace(/^https?:\\/\\//, 'x-safari-https://');
        window.location.href = safariUrl;
        
        // Fallback: try opening in new tab after 500ms
        setTimeout(function() {
          if (!document.hidden) {
            window.open(currentUrl, '_blank');
          }
        }, 500);
      } else {
        // Unknown platform - show failure immediately
        showRedirectFailed();
      }
    } catch (e) {
      console.error('PWA: Redirect error:', e);
      showRedirectFailed();
    }
  };

  // ============================================================
  // SECTION 5: Clipboard Handling
  // Copies the current URL to clipboard with modern API fallback
  // Shows visual feedback on success/failure
  // ============================================================
  window.pwaCopyLink = function() {
    var currentUrl = window.location.href;
    var btn = document.querySelector('.pwa-copy-button');
    
    // Success handler - update button text
    function onCopySuccess() {
      if (btn) {
        btn.textContent = '✓ Link Copied!';
        setTimeout(function() { btn.innerHTML = '📋 Copy Link'; }, 2000);
      }
    }
    
    // Failure handler - show error state
    function onCopyFailed() {
      if (btn) {
        btn.textContent = '✗ Copy failed';
        btn.style.color = '#ff6b6b';
        setTimeout(function() { 
          btn.innerHTML = '📋 Copy Link'; 
          btn.style.color = '';
        }, 2000);
      }
      console.error('PWA: Clipboard copy failed');
    }
    
    // Try modern Clipboard API first
    if (navigator.clipboard && navigator.clipboard.writeText) {
      navigator.clipboard.writeText(currentUrl)
        .then(onCopySuccess)
        .catch(onCopyFailed);
    } else {
      // Fallback: create temporary textarea for execCommand
      var textArea = document.createElement('textarea');
      textArea.value = currentUrl;
      textArea.style.position = 'fixed';
      textArea.style.left = '-9999px';
      document.body.appendChild(textArea);
      textArea.select();
      try {
        var success = document.execCommand('copy');
        if (success) {
          onCopySuccess();
        } else {
          onCopyFailed();
        }
      } catch (e) {
        onCopyFailed();
      }
      document.body.removeChild(textArea);
    }
  };

  // ============================================================
  // SECTION 6: Initialization
  // Runs detection and shows overlay if in-app browser detected
  // Auto-attempts redirect on detection
  // ============================================================
  var detection = detectInAppBrowser();
  if (detection.isInApp) {
    console.log('PWA: Detected in-app browser: ' + detection.browser);
    // Show overlay first, then attempt redirect
    showOverlay(detection.browser);
    // Small delay to ensure overlay is visible before redirect attempt
    setTimeout(function() {
      try { 
        if (window.pwaAttemptRedirect) window.pwaAttemptRedirect(); 
      } catch (e) { 
        console.error('PWA: Auto-redirect failed', e); 
      }
    }, 100);
  }
})();
''';
  }
}
